google/cloud/pubsub/samples/samples.cc

```diff
@@ -594,23 +601,30 @@ void SubscribeCustomAttributes(
   using google::cloud::future;
   using google::cloud::StatusOr;
   [](pubsub::Subscriber subscriber, pubsub::Subscription const& subscription) {
-    std::atomic<int> count{0};
-    google::cloud::promise<void> received_message;
-    auto result = subscriber.Subscribe(
+    std::mutex mu;
+    std::condition_variable cv;
+    bool done = false;
+    int message_count = 0;
+    auto session = subscriber.Subscribe(
         subscription, [&](pubsub::Message const& m, pubsub::AckHandler h) {
           std::cout << "Received message with attributes:\n";
           for (auto const& kv : m.attributes()) {
             std::cout << "  " << kv.first << ": " << kv.second << "\n";
           }
-          // synchronize with the waiting thread to gracefully end the example
-          if (count++ == 0) received_message.set_value();
+          std::unique_lock<std::mutex> lk(mu);
+          ++message_count;
+          done = true;
+          cv.notify_one();
+          lk.unlock();
           std::move(h).ack();
         });
-    received_message.get_future()
-        .then([&result](google::cloud::future<void>) { result.cancel(); })
-        .get();
-    std::cout << "Message count = " << count.load()
-              << ", status = " << result.get() << "\n";
+    // Most applications would just release the `session` object at this point,
+    // but we want to gracefully close down this example.
+    std::unique_lock<std::mutex> lk(mu);
+    cv.wait(lk, [&done] { return done; });
+    session.cancel();
```

Comment:
In the previous examples we were careful to release the lock before canceling the subscription.  Was that important?  If so, why not here?  If not, why there?

Reply:
Actually I was pretty inconsistent, the only case where it mattered was around line 586 (new reckoning), where the `.then()` callback would try to acquire the mutex again.  Changed all three examples to look the same.

Reply:
I ran the program under `tsan` and `asan` 2,000 times (each) to make sure nothing bad was happening.